<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Справочник</title>
  <link rel="stylesheet" href="../css/project-test.css">
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Roboto&display=swap" rel="stylesheet">
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@200&display=swap" rel="stylesheet">
  <link href="../css/prism.css" rel="stylesheet">
  <script src="../js/prism.js"></script>
  <script src="../js/project.js"></script>
</head>

<body>
  <div class="main-wrapper">
    <main id="content" class="main_content" role="main">
      <article class="main-page-content" lang="ru"> <!-- lang="ru не совсем понял для чего-->
        <section aria-labelledby="Что не так с этим тестом?">
                <h3 id="Что не так с этим тестом?">Что не так с этим тестом?</h3>
                <div class="section-content">
                  <p>Что не так в нижеприведённом тесте функции <code>pow</code>?</p>
                  <pre><code class="language-js">it("Возводит x в степень n", function() {
  let x = 5;

  let result = x;
  assert.equal(pow(x, 1), result);

  result *= x;
  assert.equal(pow(x, 2), result);

  result *= x;
  assert.equal(pow(x, 3), result);
});</code></pre>
                  <p>P.S. Тест не содержит синтаксических ошибок и успешно проходит.</p>
                  <details>
                    <summary class="solution">
                      <li class="task__solution" type="button">решение</li>
                    </summary>
                    <div>
                      <p>Тест демонстрирует один из соблазнов, с которым сталкиваются разработчики при их написании.</p>
                      <p>У нас тут, по сути, три теста, но они написаны как одна функция с тремя проверками.</p>
                      <p>Иногда так проще писать, но если произойдёт ошибка, то гораздо сложнее понять, что пошло не
                        так.
                      </p>
                      <p>Если ошибка происходит посередине сложного потока выполнения, то нам придётся выяснять, какие
                        данные были в этом месте. По сути, придётся <em>отлаживать тест</em>.</p>
                      <p>Гораздо лучше разбить тест на несколько блоков <code>it</code> и ясно описать входные и
                        ожидаемые
                        на выходе данные.</p>
                      <p>Примерно так:</p>
                      <pre><code class="language-js">describe("Возводит x в степень n", function() {
  it("5 в степени 1 будет 5", function() {
    assert.equal(pow(5, 1), 5);
  });

  it("5 в степени 2 будет 25", function() {
    assert.equal(pow(5, 2), 25);
  });

  it("5 в степени 3 будет 125", function() {
    assert.equal(pow(5, 3), 125);
  });
});</code></pre>
                      <p>Мы заменили один <code>it</code> на <code>describe</code> и группу блоков <code>it</code>.
                        Теперь, если какой-либо из блоков завершится неудачно, мы точно увидим, с какими данными это
                        произошло.</p>
                      <p>Также мы можем изолировать один тест и запускать только его, написав <code>it.only</code>
                        вместо
                        <code>it</code>:
                      </p>
                      <pre><code class="language-js">describe("Возводит x в степень n", function() {
  it("5 в степени 1 будет 5", function() {
    assert.equal(pow(5, 1), 5);
  });

  // Mocha будет запускать только этот блок
  it.only("5 в степени 2 будет 25", function() {
    assert.equal(pow(5, 2), 25);
  });

  it("5 в степени 3 будет 125", function() {
    assert.equal(pow(5, 3), 125);
  });
});</code></pre>
                    </div>
                  </details>
                </div>
              </section>
      </article>
    </main>
  </div>
</body>
